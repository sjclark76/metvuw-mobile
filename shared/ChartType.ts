export type ChartType = 'Rain' | 'Satellite' | 'Radar' | 'Upper Air'
