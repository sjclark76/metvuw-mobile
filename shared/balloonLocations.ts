export type BalloonLocationCode = '93112' | '93417' | '93844'

export type BalloonLocation =
  | 'Whenuapai'
  | 'Paraparaumu'
  | 'Invercargill'
  | 'Raoul Island'

export const balloonLocations: Record<BalloonLocationCode, BalloonLocation> = {
  '93112': 'Whenuapai',
  '93417': 'Paraparaumu',
  '93844': 'Invercargill',
}
