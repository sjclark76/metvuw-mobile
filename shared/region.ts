// type RegionCode =
//   | 'nz'
//   | 'nzni'
//   | 'nzsi'
//   | 'victoria'
//   | 'nsw'
//   | 'waussie'
//   | 'swaussie'
//   | 'queensland'
//   | 'saussie'
//   | 'seaussie'
//   | 'newcaledonia'
//   | 'fiji'
//   | 'swp'
//   | 'ocean'
//   | 'europe'
//   | 'uk'
//   | 'estonia'
//   | 'turkey'
//   | 'world'
//   | 'satlantic'
//   | 'natlantic'
//   | 'usa'
//   | 'japan'
//   | 'safrica'

export const regions = [
  { name: 'New Zealand', code: 'nz' },
  { name: 'North Island', code: 'nzni' },
  { name: 'South Island', code: 'nzsi' },
  { name: 'Victoria & Tasmania', code: 'victoria' },
  { name: 'New South Wales', code: 'nsw' },
  { name: 'Western Australia', code: 'waussie' },
  { name: 'Perth', code: 'swaussie' },
  { name: 'Queensland', code: 'queensland' },
  { name: 'South Australia', code: 'saussie' },
  { name: 'South East Australia', code: 'seaussie' },
  { name: 'New Caledonia', code: 'newcaledonia' },
  { name: 'Fiji', code: 'fiji' },
  { name: 'South West Pacific', code: 'swp' },
  { name: 'Fiji - NZ', code: 'ocean' },
  { name: 'Europe', code: 'europe' },
  { name: 'United Kingdom', code: 'uk' },
  { name: 'Estonia', code: 'estonia' },
  { name: 'Turkey', code: 'turkey' },
  { name: 'World', code: 'world' },
  { name: 'South Atlantic', code: 'satlantic' },
  { name: 'North Atlantic', code: 'natlantic' },
  { name: 'USA', code: 'usa' },
  { name: 'Japan', code: 'japan' },
  { name: 'South Africa', code: 'safrica' },
] as const

export function findRegionByCode(code: string): Region | undefined {
  return regions.find((value) => value.code === code)
}

export const nzLinks: Region[] = [
  { name: 'New Zealand', code: 'nz' },
  { name: 'North Island', code: 'nzni' },
  { name: 'South Island', code: 'nzsi' },
]
export const australiaLinks: Region[] = [
  { name: 'Victoria & Tasmania', code: 'victoria' },
  { name: 'New South Wales', code: 'nsw' },
  { name: 'Western Australia', code: 'waussie' },
  { name: 'Perth', code: 'swaussie' },
  { name: 'Queensland', code: 'queensland' },
  { name: 'South Australia', code: 'saussie' },
  { name: 'South East Australia', code: 'seaussie' },
]
export const pacificLinks: Region[] = [
  { name: 'New Caledonia', code: 'newcaledonia' },
  { name: 'Fiji', code: 'fiji' },
  { name: 'South West Pacific', code: 'swp' },
  { name: 'Fiji - NZ', code: 'ocean' },
]
export const europeLinks: Region[] = [
  { name: 'Europe', code: 'europe' },
  { name: 'United Kingdom', code: 'uk' },
  { name: 'Estonia', code: 'estonia' },
  { name: 'Turkey', code: 'turkey' },
]
export const worldLinks: Region[] = [
  { name: 'World', code: 'world' },
  { name: 'South Atlantic', code: 'satlantic' },
  { name: 'North Atlantic', code: 'natlantic' },
  { name: 'USA', code: 'usa' },
  { name: 'Japan', code: 'japan' },
  { name: 'South Africa', code: 'safrica' },
]

export type Region = typeof regions[number]
