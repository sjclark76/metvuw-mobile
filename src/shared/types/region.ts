// type RegionCode =
//   | 'nz'
//   | 'nzni'
//   | 'nzsi'
//   | 'victoria'
//   | 'nsw'
//   | 'waussie'
//   | 'swaussie'
//   | 'queensland'
//   | 'saussie'
//   | 'seaussie'
//   | 'newcaledonia'
//   | 'fiji'
//   | 'swp'
//   | 'ocean'
//   | 'europe'
//   | 'uk'
//   | 'estonia'
//   | 'turkey'
//   | 'world'
//   | 'satlantic'
//   | 'natlantic'
//   | 'usa'
//   | 'japan'
//   | 'safrica'

export const regions = [
  { name: 'New Zealand', code: 'nz' },
  { name: 'North Island', code: 'nzni' },
  { name: 'South Island', code: 'nzsi' },
  { name: 'Victoria & Tasmania', code: 'victoria' },
  { name: 'New South Wales', code: 'nsw' },
  { name: 'Western Australia', code: 'waussie' },
  { name: 'Perth', code: 'swaussie' },
  { name: 'Queensland', code: 'queensland' },
  { name: 'South Australia', code: 'saussie' },
  { name: 'South East Australia', code: 'seaussie' },
  { name: 'New Caledonia', code: 'newcaledonia' },
  { name: 'South West Pacific', code: 'swp' },
  { name: 'Europe', code: 'europe' },
  { name: 'United Kingdom', code: 'uk' },
  { name: 'Estonia', code: 'estonia' },
  { name: 'Turkey', code: 'turkey' },
  { name: 'World', code: 'world' },
  { name: 'USA', code: 'usa' },
  { name: 'Japan', code: 'japan' },
  { name: 'South Africa', code: 'safrica' },

  { code: 'ocean', name: 'New Zealand to Fiji' },
  { code: 'pacific', name: 'Pacific Islands' },
  { code: 'fiji', name: 'Fiji Islands' },
  { code: 'spacific', name: 'South Pacific' },
  { code: 'pitcairn', name: 'Pitcairn Island' },
  { code: 'natlantic', name: 'North Atlantic' },
  { code: 'satlantic', name: 'South Atlantic' },
] as const

export const getByRegionCode = (code: string) => {
  const region = findRegionByCode(code)

  if (!region) {
    throw new Error(`invalid country code code: ${code} `)
  }

  return region
}
export const findRegionByCode = (code: string): Region | undefined =>
  regions.find((value) => value.code === code)

export const nzRegions: Region[] = [
  { name: 'New Zealand', code: 'nz' },
  { name: 'North Island', code: 'nzni' },
  { name: 'South Island', code: 'nzsi' },
]
export const australiaRegions: Region[] = [
  { name: 'Victoria & Tasmania', code: 'victoria' },
  { name: 'New South Wales', code: 'nsw' },
  { name: 'Western Australia', code: 'waussie' },
  { name: 'Perth', code: 'swaussie' },
  { name: 'Queensland', code: 'queensland' },
  { name: 'South Australia', code: 'saussie' },
  { name: 'South East Australia', code: 'seaussie' },
]
export const pacificRegions: Region[] = [
  { name: 'New Caledonia', code: 'newcaledonia' },
  { code: 'fiji', name: 'Fiji Islands' },
  { name: 'South West Pacific', code: 'swp' },
  // { name: 'Fiji - NZ', code: 'ocean' },
]
export const europeRegions: Region[] = [
  { name: 'Europe', code: 'europe' },
  { name: 'United Kingdom', code: 'uk' },
  { name: 'Estonia', code: 'estonia' },
  { name: 'Turkey', code: 'turkey' },
]
export const worldRegions: Region[] = [
  { name: 'World', code: 'world' },
  { name: 'South Atlantic', code: 'satlantic' },
  { name: 'North Atlantic', code: 'natlantic' },
  { name: 'USA', code: 'usa' },
  { name: 'Japan', code: 'japan' },
  { name: 'South Africa', code: 'safrica' },
]

export const oceans: Region[] = [
  { code: 'ocean', name: 'New Zealand to Fiji' },
  { code: 'pacific', name: 'Pacific Islands' },
  { code: 'fiji', name: 'Fiji Islands' },
  { code: 'spacific', name: 'South Pacific' },
  { code: 'pitcairn', name: 'Pitcairn Island' },
  { code: 'natlantic', name: 'North Atlantic' },
  { code: 'satlantic', name: 'South Atlantic' },
]

export type Region = (typeof regions)[number]
