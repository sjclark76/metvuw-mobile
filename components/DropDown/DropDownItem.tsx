import { Region } from '../../shared/region'

interface DropDownItemProps {
  region: Region
}

export const DropDownItem = (props: DropDownItemProps) => {
  return (
    <>
      <li key={props.region.code} className="hover:bg-blue-500 rounded">
        <a
          className="max-w-full py-2 px-4 block whitespace-no-wrap "
          href={`/regions/${props.region.code}`}
        >
          {props.region.name}
        </a>
      </li>
    </>
  )
}
