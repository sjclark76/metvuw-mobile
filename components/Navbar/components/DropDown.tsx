import { MenuLink } from '../types'
import { useEffect, useRef, useState } from 'react'
import styles from '../NavBar.module.css'
interface DropDownProps {
  heading: string
  links: MenuLink[]
}
const DropDown = (props: DropDownProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [currentItem, setCurrentItem] = useState<number | undefined>(undefined)
  const wrapperRef = useRef<HTMLButtonElement>(null)

  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown)
    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  })

  const handleKeyDown = (e) => {
    if (isOpen) {
      e.preventDefault()

      switch (e.keyCode) {
        // KeyDown
        case 40:
          if (currentItem === props.links.length - 1) {
            setCurrentItem(0)
          } else {
            const newValue = currentItem === undefined ? 0 : currentItem + 1
            setCurrentItem(newValue)
          }
          break
        // KeyUp
        case 38:
          if (currentItem === 0) {
            setCurrentItem(props.links.length - 1)
          } else {
            setCurrentItem(
              currentItem ? currentItem - 1 : props.links.length - 1,
            )
          }
          break
        // Escape
        case 27:
          setCurrentItem(1)
          setIsOpen(false)
          break
        default:
          break
      }
    }
  }
  return (
    <div className="relative inline-flex " id="dropdown">
      <button
        className={`${styles.headerText}`}
        onClick={() => setIsOpen((prevState) => !prevState)}
        aria-expanded={isOpen}
        ref={wrapperRef}
      >
        <span>{props.heading}</span>
        <span className="relative only:-mx-5">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            className="w-5 h-5"
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
            strokeWidth="1.5"
            aria-labelledby="t-01 d-01"
            role="graphics-symbol"
          >
            <title id="t-01">Button icon</title>
            <desc id="d-01">An icon describing the buttons usage</desc>
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              d="M19.5 8.25l-7.5 7.5-7.5-7.5"
            />
          </svg>
        </span>
      </button>
      <ul
        className={`${
          isOpen ? 'flex' : 'hidden'
        } absolute top-full z-10 mt-3 flex w-72 list-none flex-col rounded bg-white py-2 shadow-md shadow-slate-500/10 `}
      >
        {props.links.map((item, index) => {
          return (
            <li key={index}>
              <a
                className={` ${
                  index === currentItem
                    ? 'bg-emerald-50 text-blue-500'
                    : 'bg-none text-slate-500'
                } flex items-start justify-start gap-2 p-2 px-5 transition-colors duration-300 hover:bg-blue-50 hover:text-blue-500 focus:bg-blue-50 focus:text-blue-600 focus:outline-none focus-visible:outline-none`}
                href={item.href}
                aria-current={index + 1 === currentItem ? 'page' : 'false'}
              >
                <span className="flex flex-col gap-1 overflow-hidden whitespace-nowrap">
                  <span className="truncate leading-5">{item.value}</span>
                </span>
              </a>
            </li>
          )
        })}
      </ul>
    </div>
  )
}

export default DropDown
