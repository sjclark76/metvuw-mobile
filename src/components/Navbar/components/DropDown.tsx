import { MenuLink } from '../types'
import { useEffect, useRef, useState } from 'react'
import styles from '../NavBar.module.css'
import Favourite from '../../Favourite'
interface DropDownProps {
  heading: string
  links: MenuLink[]
}
const DropDown = (props: DropDownProps) => {
  const [isOpen, setIsOpen] = useState(false)
  const [currentItem, setCurrentItem] = useState<number | undefined>(undefined)
  const menuRef = useRef<HTMLUListElement>(null)
  useEffect(() => {
    window.addEventListener('keydown', handleKeyDown)
    return () => {
      window.removeEventListener('keydown', handleKeyDown)
    }
  })

  useEffect(() => {
    function handleClickOutside(event) {
      if (menuRef.current && !menuRef.current.contains(event.target)) {
        setIsOpen(false)
      }
    }

    document.addEventListener('mousedown', handleClickOutside)

    return () => {
      document.removeEventListener('mousedown', handleClickOutside)
    }
  }, [menuRef])

  const handleKeyDown = (e: KeyboardEvent) => {
    if (isOpen) {
      e.preventDefault()

      switch (e.keyCode) {
        // KeyDown
        case 40:
          if (currentItem === props.links.length - 1) {
            setCurrentItem(0)
          } else {
            const newValue = currentItem === undefined ? 0 : currentItem + 1
            setCurrentItem(newValue)
          }
          break
        // KeyUp
        case 38:
          if (currentItem === 0) {
            setCurrentItem(props.links.length - 1)
          } else {
            setCurrentItem(
              currentItem ? currentItem - 1 : props.links.length - 1,
            )
          }
          break
        // Escape
        case 27:
          setCurrentItem(1)
          setIsOpen(false)
          break
        default:
          break
      }
    }
  }
  return (
    <div className="relative inline-flex " id="dropdown">
      <button
        className={`${styles.headerText}`}
        onClick={() => setIsOpen((prevState) => !prevState)}
        aria-expanded={isOpen}
      >
        <span>{props.heading}</span>
        <span className="relative only:-mx-5">
          <svg
            xmlns="http://www.w3.org/2000/svg"
            className="w-5 h-5"
            fill="none"
            viewBox="0 0 24 24"
            stroke="currentColor"
            strokeWidth="1.5"
            aria-labelledby="t-01 d-01"
            role="graphics-symbol"
          >
            <title id="t-01">Button icon</title>
            <desc id="d-01">An icon describing the buttons usage</desc>
            <path
              strokeLinecap="round"
              strokeLinejoin="round"
              d="M19.5 8.25l-7.5 7.5-7.5-7.5"
            />
          </svg>
        </span>
      </button>
      <ul
        ref={menuRef}
        className={`${
          isOpen ? 'flex' : 'hidden'
        } absolute top-full z-10 mt-3 flex w-72 list-none flex-col rounded bg-white py-2 shadow-md shadow-slate-500/10 `}
      >
        {props.links.map((item, index) => {
          return (
            <li
              key={index}
              className={` ${
                index === currentItem
                  ? 'bg-emerald-50 text-blue-500'
                  : 'bg-none text-slate-500'
              } flex items-start justify-between gap-2 p-2 px-5 transition-colors duration-300 hover:bg-blue-50 hover:text-blue-500 focus:bg-blue-50 focus:text-blue-600 focus:outline-none focus-visible:outline-none`}
            >
              <a
                href={item.href}
                aria-current={index + 1 === currentItem ? 'page' : 'false'}
              >
                <span className="flex flex-col gap-1 overflow-hidden whitespace-nowrap">
                  <span className="truncate leading-5">{item.value}</span>
                </span>
              </a>
              <Favourite menuLink={item} />
            </li>
          )
        })}
      </ul>
    </div>
  )
}

export default DropDown
