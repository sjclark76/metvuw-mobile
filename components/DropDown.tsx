import { DropDownItem } from './DropDownItem'
import { Region } from '../shared/region'

interface DropDownProps {
  heading: string
  links: Region[]
}

export const DropDown = (props: DropDownProps) => {
  return (
    <>
      <div className="group inline-block relative">
        <button className="px-3 py-2 rounded-t text-white font-bold items-center justify-center hover:bg-blue-400 hover:text-white inline-flex">
          <span className="mr-1">{props.heading}</span>
          <svg
            className="fill-current h-4 w-4"
            xmlns="http://www.w3.org/2000/svg"
            viewBox="0 0 20 20"
          >
            <path d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z" />
          </svg>
        </button>
        <ul className="absolute w-full	hidden rounded-b text-white font-bold bg-blue-400  group-hover:block z-50">
          {props.links.map((link) => (
            <DropDownItem key={link.code} region={link} />
          ))}
        </ul>
      </div>
    </>
  )
}
