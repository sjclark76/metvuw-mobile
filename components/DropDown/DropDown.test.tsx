import { render } from '@testing-library/react'
import DropDown from './DropDown'
import { nzLinks } from '../../shared/region'

describe('NavBar Component', () => {
  test('should render correctly', () => {
    const { container } = render(
      <DropDown heading="My Heading" links={nzLinks} />
    )

    expect(container).toMatchSnapshot()
  })
})
