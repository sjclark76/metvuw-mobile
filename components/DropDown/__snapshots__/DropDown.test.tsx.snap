// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NavBar Component should render correctly 1`] = `
<div>
  <div
    class="group inline-block relative"
  >
    <button
      class="px-3 py-2 rounded-t text-white font-bold items-center justify-center hover:bg-blue-400 hover:text-white inline-flex"
    >
      <span
        class="mr-1"
      >
        My Heading
      </span>
      <svg
        class="fill-current h-4 w-4"
        viewBox="0 0 20 20"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M9.293 12.95l.707.707L15.657 8l-1.414-1.414L10 10.828 5.757 6.586 4.343 8z"
        />
      </svg>
    </button>
    <ul
      class="absolute w-full	hidden rounded-b text-white font-bold bg-blue-400  group-hover:block z-50"
    >
      <li
        class="hover:bg-blue-500 rounded"
      >
        <a
          class="max-w-full py-2 px-4 block whitespace-no-wrap "
          href="/regions/nz"
        >
          New Zealand
        </a>
      </li>
      <li
        class="hover:bg-blue-500 rounded"
      >
        <a
          class="max-w-full py-2 px-4 block whitespace-no-wrap "
          href="/regions/nzni"
        >
          North Island
        </a>
      </li>
      <li
        class="hover:bg-blue-500 rounded"
      >
        <a
          class="max-w-full py-2 px-4 block whitespace-no-wrap "
          href="/regions/nzsi"
        >
          South Island
        </a>
      </li>
    </ul>
  </div>
</div>
`;
